---
title: "Introduction to pmsignature"
author: "Yuichi Shiraishi"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::knitr}
  %\usepackage[utf8]{inputenc}
---



The R package `pmsignature` is developed 
for efficiently extracting characteristic mutation patterns (mutation signatures) 
from the set of mutations collected typically from cancer genome sequencing data.

For extracting mutation signatures, 
principal component analysis or nonnegative matrix factorization have been popular.
Compared to these existing approaches, the `pmsignature` has following advantages:

1. `pmsignature` can perform robust estimation of mutation signatures even taking account for many mutation features such as two bases 5' and 3' to the mutated sites.
1. `pmsignature` provides intuitively interetable visualization of mutation signatures, which is reminicent of sequencing logos.

Currently, `pmsignature` can only accept tab delimited text files with specialized format. 
We will improve the program so that it can accept VCF format files.

## Paper

> Shiraishi et al. Extraction of Latent Probabilistic Mutational Signature in Cancer Genomes, submitted.

## Input data

The input data must be a tab-delimited text file.
An example of input data for `pmsignature` is as follows:

---  --  --  --  --  --  --
 26   6  
  6   4   4   4   4   2   
  1   4   4   4   3   3   2
  2   4   3   3   1   1   2
  3   4   4   3   2   2   2
  4   3   3   2   3   3   1
  5   3   4   2   4   4   2
  6   4   1   4   2   1   2
  3   2   1   1   1   1   2
  7   4   2   2   4   3   2
---  --  --  --  --  --  --

- The first row shows the number of cancer genomes and mutation features
(here, the number of cancer genome and mutation features are 26 and 6, respectively).

- The second row shows the possible number of values that each mutation feature can take.
In this example, they consider substitution patterns (6 values, C>A, C>G, C>T, T>A, T>C and T>G),
two 5' and 3' flanking bases (4 values for each 4 site, A, C, G and T)
and transcription direction (2 values, + and -).

- The third and later rows show the mutation features for each mutation. The first column shows the index of cancer genome. From the second to last columns shows the value of mutation features.

mutation feature
: elements used for categorize the mutations. 
Typically 6 substitutions (C>A, C>G, C>T, T>A, T>C and T>G), 
5' and 3' flanking bases (A, C, G and T), and transcription direction.


## Workflow

### Install the package

The easiest way is to use the package `devtools`:

```
devtools::install_github("friend1ws/pmsignature");
```



### Prepare input data

First, create the input data from your mutation data.

After installing `pmsignature`,
you can find the above example file at the directory where pmsignature is installed.

```
inputFile <- system.file("extdata", "hoang.ind5_dir.txt", package="pmsignature");
```

For converting your own mutation list data,
the [script](https://twitter.com/hadleywickham/status/504368538874703872) 
we prepared might be helpful.

We are now improving the implementation so that the package can handle 
the most typical format for mutation data is VCF format.

Also, you might want to use the background signature files.
The background data used in the original paper can be found at:

```
bgFile <- system.file("bgdata", "bg.ind5_dir.txt", package="pmsignature");
```

### Read input data

Type the following commands:
```
G <- readRawMutFile(inputFile);
```

Also, you want to read the background signature files as follow:

```
BG0 <- readBGFile(bgFile);
```

### Estimate the parameters

After setting the number of mutation signature (e.g., 3),

```
K <- 3
```

type the following command:

```
pms_param <- getPMSignature(G, K, fdim);
```

If you want to add the background signature, then type

```
pms_param <- getPMSignature(G, K, fdim, isBG=TRUE, BG0=BG0);
```

In default, we repeat the estimation 10 times by changing the initial value.
If you want to changet the repeat number, then

```
pms_param <- getPMSignature(G, K, fdim, isBG=TRUE, BG0=BG0, numInit=20);
```


### Visualing the mutation signatures and memberships

Currently, we have provided 


